File: Codes/Level 2/Recursion and Backtracking/Permutations and Combinations/Permutations2Leetcode.cpp
class Solution {
public:
    
    map<vector<int>, int> m;
    
    void permute(vector<vector<int>> &ans, vector<int> &currAns, vector<int> &nums, vector<bool> &visited){
        if(currAns.size() == nums.size()){
            if(m.find(currAns) == m.end()){
                ans.push_back(currAns);
                m[currAns]++;
            }
            return;
        }
        
        for(int i=0; i<nums.size(); i++){
            if(!visited[i]){
                visited[i] = true;
                currAns.push_back(nums[i]);
                permute(ans, currAns, nums, visited);
                currAns.pop_back();
                visited[i] = false;
            }
        }
    }
    
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> currAns;
        vector<bool> visited(nums.size());
        
        permute(ans, currAns, nums, visited);
        
        return ans;
    }
};

File: Codes/Level 2/Recursion and Backtracking/Permutations and Combinations/Permutations2Leetcode.java
class Solution {
    
    HashMap<List<Integer>, Integer> m = new HashMap<>();
    
    public void permute(List<List<Integer>> ans, List<Integer> currAns, int[] nums, boolean[] visited){
        if(currAns.size() == nums.length){
            if(m.containsKey(currAns) == false){
                List<Integer> temp = new ArrayList<>();
                for(Integer i : currAns)
                    temp.add(i);
                ans.add(temp);
                m.put(currAns, 1);
            }
            return;
        }
        
        for(int i=0; i<nums.length; i++){
            if(visited[i] == false){
                visited[i] = true;
                currAns.add(nums[i]);
                permute(ans, currAns, nums, visited);
                currAns.remove(currAns.size() - 1);
                visited[i] = false;
            }
        }
    }
    
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> currAns = new ArrayList<>();
        boolean[] visited = new boolean[nums.length];
        permute(ans, currAns, nums, visited);
        return ans;
    }
}